File: src/steam/proto.ts
// A protobuf codec with no schema: fields go in and come out by number.
// Steam's messages are read once and thrown away, so generated classes would
// cost more than they save — what matters is that the field numbers are right,
// and those are recorded in docs/STEAM-DEPOT.md.

export type Field = bigint | Uint8Array;

export class Writer {
  private parts: Uint8Array[] = [];

  private varint(value: bigint): void {
    const bytes: number[] = [];
    let v = value;
    do {
      let b = Number(v & 0x7fn);
      v >>= 7n;
      if (v !== 0n) b |= 0x80;
      bytes.push(b);
    } while (v !== 0n);
    this.parts.push(new Uint8Array(bytes));
  }

  private tag(field: number, wire: number): void {
    this.varint(BigInt((field << 3) | wire));
  }

  uint(field: number, value: number | bigint): this {
    this.tag(field, 0);
    this.varint(BigInt(value));
    return this;
  }

  bool(field: number, value: boolean): this {
    return this.uint(field, value ? 1 : 0);
  }

  fixed64(field: number, value: bigint): this {
    this.tag(field, 1);
    const buf = new Uint8Array(8);
    new DataView(buf.buffer).setBigUint64(0, value, true);
    this.parts.push(buf);
    return this;
  }

  bytes(field: number, value: Uint8Array): this {
    this.tag(field, 2);
    this.varint(BigInt(value.length));
    this.parts.push(value);
    return this;
  }

  string(field: number, value: string): this {
    return this.bytes(field, new TextEncoder().encode(value));
  }

  message(field: number, inner: Writer): this {
    return this.bytes(field, inner.finish());
  }

  finish(): Uint8Array {
    const total = this.parts.reduce((n, p) => n + p.length, 0);
    const out = new Uint8Array(total);
    let at = 0;
    for (const part of this.parts) {
      out.set(part, at);
      at += part.length;
    }
    return out;
  }
}

/** Field number to every value seen, because repeated fields are normal here. */
export type Message = Map<number, Field[]>;

export function read(buf: Uint8Array): Message {
  const out: Message = new Map();
  const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  let i = 0;

  const varint = (): bigint => {
    let value = 0n;
    let shift = 0n;
    while (i < buf.length) {
      const b = buf[i++];
      value |= BigInt(b & 0x7f) << shift;
      shift += 7n;
      if ((b & 0x80) === 0) break;
    }
    return value;
  };

  const push = (field: number, value: Field) => {
    const list = out.get(field);
    if (list) list.push(value);
    else out.set(field, [value]);
  };

  while (i < buf.length) {
    const tag = Number(varint());
    const field = tag >> 3;
    const wire = tag & 7;
    if (wire === 0) {
      push(field, varint());
    } else if (wire === 1) {
      push(field, view.getBigUint64(i, true));
      i += 8;
    } else if (wire === 2) {
      const length = Number(varint());
      push(field, buf.subarray(i, i + length));
      i += length;
    } else if (wire === 5) {
      push(field, BigInt(view.getUint32(i, true)));
      i += 4;
    } else {
      break;
    }
  }
  return out;
}

export const num = (m: Message, field: number, fallback = 0n): bigint => {
  const value = m.get(field)?.[0];
  return typeof value === "bigint" ? value : fallback;
};

export const str = (m: Message, field: number): string | null => {
  const value = m.get(field)?.[0];
  return value instanceof Uint8Array ? new TextDecoder().decode(value) : null;
};

export const raw = (m: Message, field: number): Uint8Array | null => {
  const value = m.get(field)?.[0];
  return value instanceof Uint8Array ? value : null;
};

export const list = (m: Message, field: number): Message[] =>
  (m.get(field) ?? [])
    .filter((v): v is Uint8Array => v instanceof Uint8Array)
    .map(read);

File: uwp/Kiosk/Steam/Proto.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace Kiosk.Steam
{
    /// <summary>
    /// A protobuf codec with no schema: fields go in and come out by number.
    /// Steam's messages are read once and thrown away, so generated classes
    /// would cost more than they save. The field numbers are in
    /// docs/STEAM-DEPOT.md, read from the published .proto files.
    /// </summary>
    public sealed class ProtoWriter
    {
        private readonly MemoryStream stream = new MemoryStream();

        private void Varint(ulong value)
        {
            do
            {
                var b = (byte)(value & 0x7F);
                value >>= 7;
                if (value != 0) b |= 0x80;
                stream.WriteByte(b);
            } while (value != 0);
        }

        private void Tag(int field, int wire) => Varint((ulong)((field << 3) | wire));

        public ProtoWriter Uint(int field, ulong value)
        {
            Tag(field, 0);
            Varint(value);
            return this;
        }

        public ProtoWriter Bool(int field, bool value) => Uint(field, value ? 1UL : 0UL);

        public ProtoWriter Fixed64(int field, ulong value)
        {
            Tag(field, 1);
            var bytes = BitConverter.GetBytes(value);
            stream.Write(bytes, 0, 8);
            return this;
        }

        public ProtoWriter Bytes(int field, byte[] value)
        {
            Tag(field, 2);
            Varint((ulong)value.Length);
            stream.Write(value, 0, value.Length);
            return this;
        }

        public ProtoWriter String(int field, string value) =>
            Bytes(field, Encoding.UTF8.GetBytes(value ?? string.Empty));

        public ProtoWriter Message(int field, ProtoWriter inner) => Bytes(field, inner.Finish());

        public byte[] Finish() => stream.ToArray();
    }

    /// <summary>Field number to every value seen; repeated fields are normal.</summary>
    public sealed class ProtoMessage
    {
        private readonly Dictionary<int, List<object>> fields = new Dictionary<int, List<object>>();

        public static ProtoMessage Read(byte[] buffer) => Read(buffer, 0, buffer.Length);

        public static ProtoMessage Read(byte[] buffer, int start, int length)
        {
            var message = new ProtoMessage();
            var i = start;
            var end = start + length;

            ulong Varint()
            {
                ulong value = 0;
                var shift = 0;
                while (i < end)
                {
                    var b = buffer[i++];
                    value |= (ulong)(b & 0x7F) << shift;
                    shift += 7;
                    if ((b & 0x80) == 0) break;
                }
                return value;
            }

            while (i < end)
            {
                var tag = Varint();
                var field = (int)(tag >> 3);
                var wire = (int)(tag & 7);
                if (wire == 0)
                {
                    message.Add(field, Varint());
                }
                else if (wire == 1)
                {
                    message.Add(field, BitConverter.ToUInt64(buffer, i));
                    i += 8;
                }
                else if (wire == 2)
                {
                    var size = (int)Varint();
                    var data = new byte[size];
                    Array.Copy(buffer, i, data, 0, size);
                    i += size;
                    message.Add(field, data);
                }
                else if (wire == 5)
                {
                    message.Add(field, (ulong)BitConverter.ToUInt32(buffer, i));
                    i += 4;
                }
                else
                {
                    break;
                }
            }
            return message;
        }

        private void Add(int field, object value)
        {
            if (!fields.TryGetValue(field, out var list))
            {
                list = new List<object>(1);
                fields[field] = list;
            }
            list.Add(value);
        }

        public ulong Num(int field, ulong fallback = 0)
        {
            if (fields.TryGetValue(field, out var list) && list.Count > 0 && list[0] is ulong value)
            {
                return value;
            }
            return fallback;
        }

        public byte[] Raw(int field)
        {
            if (fields.TryGetValue(field, out var list) && list.Count > 0 && list[0] is byte[] value)
            {
                return value;
            }
            return null;
        }

        public string Str(int field)
        {
            var raw = Raw(field);
            return raw == null ? null : Encoding.UTF8.GetString(raw, 0, raw.Length);
        }

        public List<ProtoMessage> List(int field)
        {
            var out_ = new List<ProtoMessage>();
            if (!fields.TryGetValue(field, out var list)) return out_;
            foreach (var value in list)
            {
                if (value is byte[] bytes) out_.Add(Read(bytes));
            }
            return out_;
        }

        public bool Has(int field) => fields.ContainsKey(field);
    }
}
